tablets: Add support for removenode and replace handling

tgrabiec opened the pull request:
New tablet replicas are allocated and rebuilt synchronously with node
operations. They are safely rebuilt from all existing replicas.
The list of ignored nodes passed to node operations is respected.

Tablet scheduler is responsible for scheduling tablet rebuilding transition which
changes the replicas set. The infrastructure for handling decommission
in tablet scheduler is reused for this.

Scheduling is done incrementally, respecting per-shard load
limits. Rebuilding transitions are recognized by load calculation to
affect all tablet replicas.

New kind of tablet transition is introduced called "rebuild" which
adds new tablet replica and rebuilds it from existing replicas. Other
than that, the transition goes through the same stages as regular
migration to ensure safe synchronization with request coordinators.

In this PR we simply stream from all tablet replicas. Later we should
switch to calling repair to avoid sending excessive amounts of data.

Fixes https://github.com/scylladb/scylladb/issues/16690.

tgrabiec commented:
Please ignore the first two commits, they come from PR https://github.com/scylladb/scylladb/pull/14995

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1967,8 +1967,16 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                if (f.failed()) {
+                    auto ep = f.get_exception();
+                    slogger.error("raft topology: {} for tablet {} failed: {}", name, gid, ep);
+                    return seastar::sleep_abortable(std::chrono::seconds(1), _as).then([ep] () mutable {
```
Why not `co_await`?

avikivity replied:
Would fall into the lambda coroutine fiasco.

gleb-cloudius replied:
The only thing we need after `co_await` is ep which is a local var.

avikivity commented on docs/dev/topology-over-raft.md:
```diff
@@ -104,19 +104,28 @@ that there are no tablet transitions in the system.
+ * migration - tablet replica is moved from one shard to another (possibly on a different node)
+
+ * rebuild - new tablet replica is rebuilt from existing ones, possibly dropping old replica afterwards (on node removal or replace)
+
```
In the future, we may want to make operations in different datacenters  (or even racks) independent. So we can migrate a tablet in dc1, and rebuild it in dc2. This won't work with this representation.

I don't propose fixing it now. I guess we could have had a separate transitions table to list individual transitions with better than tablet granularity. At least with raft changing schemas atomically is easier.

avikivity commented on service/storage_service.cc:
```diff
@@ -1964,6 +1964,9 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
         if (!holder || holder->failed()) {
+            if (holder && holder->failed()) {
+                holder->ignore_ready_future(); // Prevent warnings in logs
+            }
```
Needs comment to explain why it's okay to prevent warnings.

avikivity commented on service/storage_service.cc:
```diff
@@ -1967,8 +1967,16 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                if (f.failed()) {
+                    auto ep = f.get_exception();
+                    slogger.error("raft topology: {} for tablet {} failed: {}", name, gid, ep);
+                    return seastar::sleep_abortable(std::chrono::seconds(1), _as).then([ep] () mutable {
```
Suggest exponential backoff.

tgrabiec replied:
100ms -> 4s ?

asias replied:
Will it retry forever? How can the user abort the ops? After the abort, will it rollback to a previous state as if removenode/replace did not happen?

tgrabiec replied:
It will stop retrying when the operation eventually succeeds.

There is no support for automatic abort.

There will be rollback if the involved node is dead and admin marks it as excluded: https://github.com/scylladb/scylladb/issues/16527

avikivity commented on service/storage_service.cc:
```diff
@@ -1967,8 +1967,16 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                    return seastar::sleep_abortable(std::chrono::seconds(1), _as).then([ep] () mutable {
+                        std::rethrow_exception(ep);
+                    });
+                }
```
Overall it would be better to have a fiber own the work, rather than managing random futures. But that's outside the scope here.

avikivity commented on locator/tablets.cc:
```diff
@@ -84,17 +84,24 @@ tablet_migration_streaming_info get_migration_streaming_info(const tablet_info&
+        case tablet_transition_kind::rebuild:
+            result.written_to = std::unordered_set<tablet_replica>(trinfo.next.begin(), trinfo.next.end());
+            result.read_from = result.written_to;
+            return result;
```
I don't understand this at all. Don't we want to stream to just the new tablet replica? And stream from all tablets, excluding the one on the dead nodes, and the new tablet?

tgrabiec replied:
Note that this is for the purpose of load impact tracking only.

This is anticipating that we will use repair, which may write to other replicas too.

avikivity replied:
Maybe we should have separate fields for that. I really liked the consolidation where this function identifies sources and targets, and then the rest of the code doesn't care about the transition kind. Not critical for merging, though, but comments are needed to explain the unexpected computation.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -518,6 +518,12 @@ future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tm
+        if (n) {
+            n->set_excluded(true);
+        }
+    }
```
So what happens after current operation completes and next one starts. We stop marking those nodes as excluded again. So it looks like we use locator as a way to pass current topology operation parameters somewhere. This somewhere looks to be the tablet load balancer. Can it get raft topology object directly instead?

tgrabiec replied:
> So what happens after current operation completes and next one starts. We stop marking those nodes as excluded again. 

> Can it get raft topology object directly instead?

It's cleaner to have one mechanism for propagating topology snapshot, so I think existing users of token_metadata_ptr should keep using it and the long term solution is to move raft topology to locator topology, not propagating two kinds of topologies everywhere.

tgrabiec replied:
> So what happens after current operation completes and next one starts. We stop marking those nodes as excluded again.

Yes. That's why I wanted excluded nodes to be permanent, but that's left for https://github.com/scylladb/scylladb/issues/16527

It shouldn't cause problems in the happy path, since nodes which are being removed/replaced will be removed from topology too when operation ends. Nodes which are marked as ignored will be unignored and that may block migrations. That could happen even before we started topology operation which ignored them, so not a new problem.

tgrabiec replied:
If removenode fails and is aborted, nodes will be unexcluded too. This also falls into the same scenario and will be dealt with as part of https://github.com/scylladb/scylladb/issues/16527.

gleb-cloudius replied:
> > Can it get raft topology object directly instead?
> 
> It's cleaner to have one mechanism for propagating topology snapshot, so I think existing users of token_metadata_ptr should keep using it and the long term solution is to move raft topology to locator topology, not propagating two kinds of topologies everywhere.

I also think that we should not have th same information in mutiple places, but in my view in the end token_metadata should be responsible for data placement only. Everything else should be in the topology.

tgrabiec replied:
> I also think that we should not have th same information in mutiple places, but in my view in the end token_metadata should be responsible for data placement only. Everything else should be in the topology.

This patch is putting things to locator::topology, which is the right place. It's just that currently topology is carried by token_metadata_ptr, which is the "root" of the topology state. Maybe long term we should do some renaming there, and make it topology_ptr and have vnode part be a member of it.

avikivity commented on service/storage_service.cc:
```diff
@@ -7487,21 +7495,59 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+                    }
+                }
+                streamer->add_rx_ranges(table.schema()->ks_name(), std::move(ranges_per_endpoint));
+                co_await streamer->stream_async();
```
Can't we fold these two together? Doesn't get_migration_streaming_info() abstract all the differences between repair and rebuild?

tgrabiec replied:
They would have to be unfolded soon when we do repair instead.

avikivity replied:
It could still be common code. We pick one of the set to stream from, then repair if the set size was larger then one. For migration, that doesn't happen. Also true for rebuild with RF=2, so common code here is more general.

asias commented on service/topology_state_machine.hh:
```diff
@@ -177,6 +177,15 @@ struct topology {
     // Returns false iff we can safely start a new topology change.
     bool is_busy() const;
 
+    // Returns the set of nodes currently excluded from synchronization-with in the topology.
```
Could you add more words explaining what kind of nodes are excluded exactly,  e.g, node to be removed, node to be replaced, node are ignored by user request because a node is down?

asias replied:
But why and how is the node excluded? By who?

tgrabiec replied:
It's excluded because the node is dead and not coming back. 

It's marked in places which detect this condition, e.g. removenode, replace, or in the future, explicitly by the admin.

asias commented on locator/tablets.cc:
```diff
@@ -66,29 +66,42 @@ read_replica_set_selector get_selector_for_reads(tablet_transition_stage stage)
+            return result;
+        case tablet_transition_kind::rebuild:
+            result.written_to = std::unordered_set<tablet_replica>(trinfo.next.begin(), trinfo.next.end());
+            result.read_from = result.written_to;
```
I think this is inaccurate. If we consider using repair for the tablet replica
'rebuild', not all new replicas are written to. Only the pending node is being
written to. And the pending node, which has no data, is not being read from.

tgrabiec replied:
Yes, it's conservative (reserves more than necessary). 

If we use repair, not all replicas may be written too, but we don't know which ones will be. So prefer to avoid overload.

tgrabiec replied:
I'll make it more accurate.

asias commented on service/storage_service.cc:
```diff
@@ -7493,20 +7495,59 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+                                                                    topo_guard,
+                                                                    std::move(tables));
+                tm = nullptr;
+                streamer->add_source_filter(std::make_unique<dht::range_streamer::failure_detector_source_filter>(
```
I think we do not need this gossip down node filter anymore. If the chosen to sync nodes are down, we should fail the request and retry again. If the node is down and the user indeed want to ignore it, the user should use ignore dead option explicitly.

Same for the existing locator::tablet_transition_kind::migration case.

asias replied:
No. The filter will not be used if the new streamer->add_rx_ranges() interface is used, which provides the address to stream from directly. The filter is used for the old streamer->add_ranges() interface which selects which nodes to stream from.  We should not use the filter here.

tgrabiec replied:
Ok.

asias commented on service/storage_service.cc:
```diff
@@ -7493,20 +7495,59 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
             throw std::runtime_error(format("Cannot stream within the same node, tablet: {}, shard {} -> {}",
                                             tablet, leaving_replica.shard, trinfo->pending_replica.shard));
         }
```
```
        locator::tablet_replica leaving_replica = locator::get_leaving_replica(tinfo, *trinfo);
        if (leaving_replica.host == tm->get_my_id()) {
            // The algorithm doesn't work with tablet migration within the same node because
            // it assumes there is only one tablet replica, picked by the sharder, on local node.
            throw std::runtime_error(format("Cannot stream within the same node, tablet: {}, shard {} -> {}",
                                            tablet, leaving_replica.shard, trinfo->pending_replica.shard));
        }
```
This check should be in the locator::tablet_transition_kind::migration block.
The leaving node in replace and removenode case will not be the target of the
storage_service::stream_tablet rpc at all.

asias replied:
With regular migration,  the node receives stream_tablet rpc IS the target of migration, no?  It pulls data from other replica node. What's why we check leaving_replica.host (the src node) should not be the node itself (the dst node which pulls data from the src node), which checks src of migration != dst of migration. 

With replace and removenode, the node receives  stream_tablet rpc is the target of migration (rebuild migration kind, which pulls data from existing replicas). The leaving_replica is the dead node (to be replaced or removed). 
There is no point to check leaving_replica != the node which pulls the data and builds the replica because the code below never streams data from the leaving_replica at all. If we really want to check stream src != dst node, we should check the `host2ip(r.host)` below != the node itself, not leaving_replica for the locator::tablet_transition_kind::rebuild case. 

```
                for (auto r : trinfo->next) {
                    if (r.host != trinfo->pending_replica.host && !excluded.contains(raft::server_id(r.host.uuid()))) {
                        ranges_per_endpoint[host2ip(r.host)].emplace_back(range);
                    }
                }
```

tgrabiec replied:
I see what you mean. But the check for leaving.host != pending.host is not guarding streaming, it's guarding replication change. We don't support change in replication which involves the same node (yet). That's intra-node tablet migration. So the invariant should hold for all tablet transitions, regardless of streaming. We could stream nothing and it should still hold.

asias commented on service/storage_service.cc:
```diff
@@ -7493,20 +7495,59 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+
+                std::unordered_map<inet_address, dht::token_range_vector> ranges_per_endpoint;
+                for (auto r : trinfo->next) {
+                    if (r.host != trinfo->pending_replica.host && !excluded.contains(raft::server_id(r.host.uuid()))) {
```
Btw, the exlucded nodes (e.g., node to be replaced or removed) will not be in
trinfo->next (new replicas) in the first place. We are checking
excluded.contains only for the nodes that are ignored by user requests.

tgrabiec replied:
yes, but note that the ignored nodes are also called excluded.

asias replied:
What's why in the other thread I suggested you explaining what are the excluded nodes exactly.  It is unclear what nodes are excluded for what reason.

asias commented on test/topology_experimental_raft/test_tablets_removenode.py:
```diff
@@ -0,0 +1,195 @@
+                        " AND tablets = {'initial': 32};")
+    await cql.run_async("CREATE TABLE test2.test (pk int PRIMARY KEY, c int);")
+
+    # RF=3
```
Can we have a helper? This block  duplicated 6 times.

tgrabiec replied:
We can.

tgrabiec commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5990/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5990/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7321/)
> 
> #### Failed Tests (2/23867):
> * [test_simple_add_new_node_while_schema_changes_with_stream](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5990/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_stream) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_stream)
> * [test_simple_add_new_node_while_schema_changes_with_stream](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5990/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_stream_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_stream)

This is a known issue: https://github.com/scylladb/scylladb/issues/15370

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6006/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6006/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7336/)
#### Build Details:
- Duration: 1 hr 38 min
- Builder: spider3.cloudius-systems.com

avikivity commented:
Changes?

tgrabiec commented:
In V2:
  - Dropped 2 first commits which belong to a different PR 
  - Fixed typos in the comment about how leaving replica calculation in tablet_transition_kind
  - Improved comment about ignoring failed future in advance_in_background(): "// Prevent warnings about abandoned failed future. Logged below." 
  - Made get_migration_streaming_info() precise for rebuild, .read_from reflects excluded nodes and .written_to has only pending replica
  - Unified rebuild and migration handling in RPC handler for tablet streaming stage
  - Extracted create_keyspace() in test_tablets_removenode.py

avikivity commented:
Merging (large conflict due to moved file)
